### PR TITLE
Axis titles for Cartesian Charts

### PR DESCRIPTION
**Pull request by srmukher**
## New Behavior

Added axis titles (both x and y axis titles) which is an optional parameter for cartesian charts.
The following changes have been done:
1. The margins (left, right and bottom) have been doubled to accommodate the axis titles. Top margin di not require any modification.
2. Tooltip for both the titles have been added.
3. Titles have been truncated if it exceeds the maximum allowed space.
4. The RTL behaviour is tested.
5. The accessibility test for narrator is verified for the titles.
6. The dark mode theme is also verified.

Component tests for the axis titles could not be written as it has a dependency on a browser function 'getComputedTextLength'. Thus it can only be tested manually/End-to-End.

Showing the x and y axis titles:
image

image

With tooltip on x-axis title:
image

With tooltip on y-axis title:
image

With secondary y-axis enabled:

image

With RTL enabled:
image

With Dark mode enabled:
image

## Related Issue(s)

Work item: [7490 ](https://uifabric.visualstudio.com/iss/_workitems/edit/7490/)

### CODE REVIEW

File: change/@fluentui-react-charting-93e38689-8443-4580-bcb0-2114cf360c8f.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Adding optional axis titles",
+  "packageName": "@fluentui/react-charting",
```
Comment (AtishayMsft) [inline change/@fluentui-react-charting-93e38689-8443-4580-bcb0-2114cf360c8f.json:4]: There are a couple of errors in the PR build

File: packages/react-charting/src/components/AreaChart/AreaChart.types.ts [hunk 2]
```diff
@@ -59,6 +59,18 @@ export interface IAreaChartProps extends ICartesianChartProps {
    * Optimize area chart rendering for large data set.
    */
   optimizeLargeData?: boolean;
+
+  /**
+   * Prop to set the x axis title
+   * @default undefined
+   */
+  xAxisTitle?: string;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/AreaChart/AreaChart.types.ts:67]: These should be defined in CartesianChartProps only. That should cover all cartesian based charts.

File: packages/react-charting/src/components/HorizontalBarChart/HorizontalBarChart.types.ts [hunk 3]
```diff
@@ -88,6 +88,18 @@ export interface IHorizontalBarChartProps {
...
+   * Prop to set the x axis title
+   * @default undefined
+   */
+  xAxisTitle?: string;
+
+  /**
+   * Prop to set the y axis title
+   * @default undefined
+   */
+  yAxisTitle?: string;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/HorizontalBarChart/HorizontalBarChart.types.ts:102]: Horizontal chart type does not have a y axis.

File: packages/react-charting/src/components/SankeyChart/SankeyChart.types.ts [hunk 4]
```diff
@@ -59,6 +59,18 @@ export interface ISankeyChartProps {
...
+   * Prop to set the x axis title
+   * @default undefined
+   */
+  xAxisTitle?: string;
+
+  /**
+   * Prop to set the y axis title
+   * @default undefined
+   */
+  yAxisTitle?: string;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/SankeyChart/SankeyChart.types.ts:73]: Sankey does not have a y axis.

File: packages/react-charting/src/components/Sparkline/Sparkline.types.ts [hunk 5]
```diff
@@ -53,6 +53,18 @@ export interface ISparklineProps {
    * Call to provide customized styling that will layer on top of the variant rules.
    */
   styles?: IStyleFunctionOrObject<ISparklineStyleProps, ISparklineStyles>;
+
+  /**
+   * Prop to set the x axis title
+   * @default undefined
+   */
+  xAxisTitle?: string;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/Sparkline/Sparkline.types.ts:61]: Sparkline does not need x/y axis title

File: packages/react-charting/src/utilities/SVGTooltipText.tsx [hunk 6]
```diff
@@ -185,6 +184,7 @@ export class SVGTooltipText
 
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   private _onTooltipMouseEnter = (ev: any): void => {
+    console.log('entered _onTooltipMouseEnter');
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/SVGTooltipText.tsx:187]: revert change in file

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts [hunk 7]
```diff
@@ -15,13 +15,22 @@ export const getStyles = (props: ICartesianChartStyleProps): ICartesianChartStyl
...
-        overflow: 'hidden',
       },
       className,
     ],
     chartWrapper: {
       overflow: 'auto',
     },
+    tooltipTitle: {
+      textAlign: 'center',
+      fontFamily: 'Segoe UI',
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts:26]: fontFamily is inherited from theme. Need not define here

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts [hunk 8]
```diff
@@ -15,13 +15,22 @@ export const getStyles = (props: ICartesianChartStyleProps): ICartesianChartStyl
...
+      fontSize: '10px',
+      fontWeight: '600',
+      fontStyle: 'normal',
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts:29]: There are tokens for all these properties. Dont set hardcoded values

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts [hunk 9]
```diff
@@ -15,13 +15,22 @@ export const getStyles = (props: ICartesianChartStyleProps): ICartesianChartStyl
...
+      lineHeight: '14px',
+      color: 'var(--neutrals-web-gray-160323130, #323130)',
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts:31]: There should be token for this color

File: packages/react-charting/src/utilities/SVGTooltipText.tsx (hunk 6 above)
Comment (srmukher) [reply packages/react-charting/src/utilities/SVGTooltipText.tsx:187]: done

File: packages/react-charting/src/utilities/utilities.ts [hunk 10]
```diff
@@ -68,22 +69,22 @@ export interface IAxisData {
 export interface IMargins {
   /**
    * left margin for the chart.
-   * @default 40
+   * @default 80
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:72]: Dont change default margins. Adjust them if the title is defined.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 11]
```diff
@@ -458,6 +481,21 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
...
+                  : this.margins.left! / 2 + this.state.startFromX - this.titleMargin,
+                textAnchor: 'middle',
+                transform: `translate(0,
+                   ${svgDimensions.height - this.margins.bottom! - this.margins.top! - this.titleMargin})rotate(-90)`,
+                className: this._classNames.tooltipTitle!,
+              }}
+              maxWidth={yAxisTitleMaximumAllowedHeight}
+              wrapContent={wrapContent}
+            />
             {this.props.secondaryYScaleOptions && (
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:499]: Add title for secondary y axis also

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 12]
```diff
@@ -458,6 +481,21 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
               }, 0)`}
               className={this._classNames.yAxis}
             />
+            <SVGTooltipText
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:484]: How much is the x and y axis shifting if both x and y axis are present.

Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:484]: for yaxis title = 8px from left + 8px from right + to render the title
for xaxis title = 8px from bottom + 8px from top + to render the title

File: packages/react-charting/src/components/Sparkline/Sparkline.types.ts (hunk 5 above)
Comment (srmukher) [reply packages/react-charting/src/components/Sparkline/Sparkline.types.ts:61]: removed and kept only in cartesian base

File: packages/react-charting/src/components/SankeyChart/SankeyChart.types.ts (hunk 4 above)
Comment (srmukher) [reply packages/react-charting/src/components/SankeyChart/SankeyChart.types.ts:73]: removed and kept only in cartesian base

File: packages/react-charting/src/components/HorizontalBarChart/HorizontalBarChart.types.ts (hunk 3 above)
Comment (srmukher) [reply packages/react-charting/src/components/HorizontalBarChart/HorizontalBarChart.types.ts:102]: removed and kept only in cartesian base

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 11 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:499]: done:
image

File: packages/react-charting/src/utilities/utilities.ts (hunk 10 above)
Comment (srmukher) [reply packages/react-charting/src/utilities/utilities.ts:72]: done

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts (hunk 7 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts:26]: done

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts (hunk 9 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts]: added

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts (hunk 8 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts]: added

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx [hunk 13]
```diff
@@ -197,6 +197,9 @@ export class VerticalBarChartBasicExample extends React.Component<IVerticalBarCh
             culture={window.navigator.language}
             chartTitle="Vertical bar chart basic example "
             data={points}
+            xAxisTitle="Huge x axis title which is very long and will be truncated with ellipsis at the end
+            of the text and will be shown on hover of the text"
+            yAxisTitle="Huge y axis title which is very long and will be truncated with ellipsis"
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx:202]: Add a new example with moderate length titles. Do not add showXAxisLablesTooltip to it. Revert the change in basic example

File: packages/react-charting/src/components/GaugeChart/GaugeChart.base.tsx [hunk 14]
```diff
@@ -220,8 +220,6 @@ export class GaugeChartBase extends React.Component<IGaugeChartProps, IGaugeChar
                       : `${((this._sweepFraction[0] / this._sweepFraction[1]) * 100).toFixed()}%`
                   }
                   textProps={{
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/GaugeChart/GaugeChart.base.tsx:222]: what is this change

File: packages/react-charting/src/components/LineChart/LineChart.types.ts [hunk 15]
```diff
@@ -78,7 +78,7 @@ export interface ILineChartProps extends ICartesianChartProps {
   culture?: string;
 
   /**
-   * @default false
+   * * @default false
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/LineChart/LineChart.types.ts:81]: revert change

File: packages/react-charting/src/utilities/utilities.ts (hunk 10 above)
Comment (AtishayMsft) [reply packages/react-charting/src/utilities/utilities.ts:71]: This is still not fixed

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts [hunk 16]
```diff
@@ -15,13 +15,21 @@ export const getStyles = (props: ICartesianChartStyleProps): ICartesianChartStyl
...
-        overflow: 'hidden',
       },
       className,
     ],
     chartWrapper: {
       overflow: 'auto',
     },
+    tooltipTitle: {
+      textAlign: 'center',
+      fontSize: FontSizes.xSmall,
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts:26]: can we use theme.fonts._ here

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 17]
```diff
@@ -446,6 +466,16 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
               })`}
               className={this._classNames.xAxis}
             />
+            <SVGTooltipText
+              content={this.props.xAxisTitle}
+              textProps={{
+                x: this.margins.left! + this.state.startFromX,
+                y: svgDimensions.height - this.titleMargin,
+                className: this._classNames.tooltipTitle!,
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:474]: call this class axisTitle

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts [hunk 18]
```diff
@@ -411,6 +429,10 @@ export interface IModifiedCartesianChartProps extends ICartesianChartProps {
    */
   chartTitle?: string;
 
+  xAxisTitle?: string;
```
[the remark this answers is not in the document]
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts:432]: yes, otherwise I am unable to access these title props while rendering the component as tested in the examples file

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 19]
```diff
@@ -459,18 +489,52 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
...
+                    this.yAxisElementSecondary = e;
+                  }}
+                  id={`yAxisGElementSecondary${this.idForGraph}`}
+                  transform={`translate(${
+                    this._isRtl ? this.margins.left! : svgDimensions.width - this.margins.right!
+                  }, 0)`}
+                  className={this._classNames.yAxis}
+                />
+                <SVGTooltipText
+                  content={this.props.yAxisTitle}
```
[the remark this answers is not in the document]
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:504]: ohkay, introducing another prop

File: packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts (hunk 16 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.styles.ts:26]: done

File: packages/react-charting/src/components/LineChart/LineChart.types.ts (hunk 15 above)
Comment (srmukher) [reply packages/react-charting/src/components/LineChart/LineChart.types.ts:81]: done

File: packages/react-charting/src/components/GaugeChart/GaugeChart.base.tsx (hunk 14 above)
Comment (srmukher) [reply packages/react-charting/src/components/GaugeChart/GaugeChart.base.tsx:222]: removed

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx (hunk 13 above)
Comment (srmukher) [reply packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx:202]: done

File: packages/react-charting/src/components/AreaChart/__snapshots__/AreaChart.test.tsx.snap [hunk 20]
```diff
@@ -994,7 +1086,6 @@ exports[`AreaChart - mouse events Should render customized callout per stack on
         font-size: 14px;
         font-weight: 400;
         height: 100%;
-        overflow: hidden;
         width: 100%;
```
[the remark this answers is not in the document]
Comment (srmukher) [reply packages/react-charting/src/components/AreaChart/__snapshots__/AreaChart.test.tsx.snap]: this style was introduced in the cartesian base earlier which is no longer required. Removed it from there

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx [hunk 21]
```diff
@@ -210,6 +210,7 @@ export class VerticalBarChartBasicExample extends React.Component<IVerticalBarCh
               ) => (props ? defaultRender(props) : null),
             })}
             hideLabels={this.state.hideLabels}
+            showXAxisLablesTooltip={true}
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx:213]: revert change

Comment (srmukher) [reply packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.Basic.Example.tsx:213]: done

**Comment by fabricteam** [comment]
## 🕵 fluentuiv8 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 22]
```diff
@@ -455,6 +482,17 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
               })`}
               className={this._classNames.xAxis}
             />
+            <SVGTooltipText
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:485]: enable this SVGTooltipText element only if axis title is defined. 
That would mean no existing test snapshot will change

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 23]
```diff
@@ -115,6 +119,21 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
           : 20
         : this.props.margins?.left ?? 40,
     };
+    if (this.props.xAxisTitle !== undefined && this.props.xAxisTitle !== '') {
+      this.margins.bottom! = this.props.margins?.bottom ?? 70;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:123]: shouldn't this be margin.bottom + some X value

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 24]
```diff
@@ -115,6 +119,21 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
...
+    }
+    if (this.props.yAxisTitle !== undefined && this.props.yAxisTitle !== '') {
+      this.margins.left! = this._isRtl
+        ? this.props.margins?.right ?? this.props?.secondaryYScaleOptions
+          ? 80
+          : 40
+        : this.props.margins?.left ?? 80;
+      this.margins.right! = this._isRtl
+        ? this.props.margins?.left ?? 80
+        : this.props.margins?.right ?? this.props?.secondaryYScaleOptions
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:133]: This should be secondaryYAxistitle not secondaryYScaleOptions

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 23 above)
Comment (AtishayMsft) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:123]: same for y and secondary y as well.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts [hunk 25]
```diff
@@ -380,6 +385,25 @@ export interface ICartesianChartProps {
...
+
+  /**
+   * Prop to set the y axis title
+   * @default undefined
+   */
+  yAxisTitle?: string;
+
+  /**
+   * Prop to set the y axis title
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts:403]: update comment

Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts:403]: done

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 24 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:133]: done

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 23 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:123]: kept it fixed values to ensure we do not add up more values on top of the existing ones leading to erroneous values.

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 22 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:485]: done

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.AxisTitle.Example.tsx [hunk 26]
```diff
@@ -0,0 +1,220 @@
+import * as React from 'react';
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.AxisTitle.Example.tsx:1]: This example isn't showing in the demo site

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts [hunk 27]
```diff
@@ -380,6 +385,29 @@ export interface ICartesianChartProps {
...
+   * @default undefined
+   * Minimum bottom margin required for x axis title is 55px
+   */
+
+  xAxisTitle?: string;
+
+  /**
+   * Prop to set the y axis title
+   * @default undefined
+   * Mimimum left margin required for y axis title is 60px and for RTL is 40px
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts:400]: Fix spelling mistake

**Review by AtishayMsft — APPROVED** [review]

File: packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts (hunk 27 above)
Comment (srmukher) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.types.ts]: fixed

File: packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.AxisTitle.Example.tsx (hunk 26 above)
Comment (srmukher) [reply packages/react-examples/src/react-charting/VerticalBarChart/VerticalBarChart.AxisTitle.Example.tsx:1]: now showing
